server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/git/GitRepositoryV2.java

```diff
@@ -0,0 +1,850 @@
+        return primaryRepo.listCommits(ALL_PATH, 1, range).get(0);
+    }
+
+    /**
+     * This repository removes the old commits by creating the secondary repository that commits are pushed
+     * together with the primary repository and removing the primary repository if the secondary repository
+     * contains at least the number of {@code minRetentionCommits} and contains the recent
+     * {@code minRetentionDays} commits.
+     */
+    @Override
+    public void removeOldCommits(int minRetentionCommits, int minRetentionDays) {
+        // TODO(minwoox): provide a way to set different minRetentionCommits and minRetentionDays
+        //                in each repository.
+        if (minRetentionCommits == 0 && minRetentionDays == 0) {
+            // Not enabled.
+            return;
+        }
+
+        final InternalRepository secondaryRepo = this.secondaryRepo;
+        if (secondaryRepo != null) {
+            if (exceedsMinRetention(secondaryRepo, minRetentionCommits, minRetentionDays)) {
+                promoteSecondaryRepo(secondaryRepo);
+            }
+        } else if (exceedsMinRetention(primaryRepo, minRetentionCommits, minRetentionDays)) {
+            createSecondaryRepo();
+        }
+    }
+
+    private boolean exceedsMinRetention(InternalRepository repo,
+                                        int minRetentionCommits, int minRetentionDays) {
+        final CommitIdDatabase commitIdDatabase = repo.commitIdDatabase();
+        final Revision headRevision = repo.headRevision();
+        final Revision firstRevision = repo.firstRevision();
+        if (minRetentionCommits != 0) {
+            if (headRevision.major() - firstRevision.major() <= minRetentionCommits) {
+                return false;
+            }
+        }
+
+        if (minRetentionDays != 0) {
+            final Instant secondCommitCreationTime = repo.secondCommitCreationTimeInstant();
+            return secondCommitCreationTime != null &&
+                   Instant.now().minus(Duration.ofDays(minRetentionDays)).isBefore(secondCommitCreationTime);
+        }
+        return true;
+    }
+
+    private void promoteSecondaryRepo(InternalRepository secondaryRepo) {
+        writeLock();
+        try {
+            logger.info("Promoting the secondary repository in {}/{}.", parent.name(), originalRepoName);
+            repoMetadata.setPrimaryRepoDir(secondaryRepo.jGitRepo().getDirectory());
+            final InternalRepository primaryRepo = this.primaryRepo;
+            this.primaryRepo = secondaryRepo;
+            this.secondaryRepo = null;
+            repositoryWorker.execute(() -> {
+                closeInternalRepository(primaryRepo);
+                final File repoDir = primaryRepo.repoDir();
+                try {
+                    if (!deleteDirectory(repoDir)) {
+                        logger.warn("Failed to delete the old primary repository: {}", repoDir);
+                    }
```

Comment:
How about 1) renaming the old repo with `repo.removed` and 2) letting the old repos be removed by `PurgeSchedulingService` after 7 days.